Server coding and design guidelines

krusche opened the pull request:
### Motivation and Context
Improve the consistency of the development.

### Description
I added a first draft of server coding and design guidelines.
(The client coding and design guidelines will follow in another PR)

### Steps for Testing
1. Read and review the new documentation
2. Provide feedback if something should be changed or something is missing

derLalla commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+4. Duplication:
+===============
+
+Avoid code duplication. If we cannot reuse a method in other place then probably this method is bad and we should consider a better way to write this method. Use Abstraction to abstract common things in one place.
```
```suggestion
Avoid code duplication. If we cannot reuse a method elsewhere, then the method is probably bad and we should consider a better way to write this method. Use Abstraction to abstract common things in one place.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+5. Variables and methods declaration:
+=================================
+
+* Encapsulate the code you feel might changed in future.
```
```suggestion
* Encapsulate the code you feel might change in future.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+7. Database:
+============
+
+* Write performant queries that can also deal with more 1000 objects in a reasonable time
```
```suggestion
* Write performant queries that can also deal with more 1000 objects in a reasonable time.
```

derLalla reviewed (changes requested):
All in all good changes! But I've found a sentence which could be optimized

sjagla27 commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+6. Structure your code correctly:
+=================================
+
+* Default packages not allowed. It can cause particular problems for Spring Boot applications that use the ``@ComponentScan``, ``@EntityScan`` or ``@SpringBootApplication`` annotations since every class from every jar is read.
```
```suggestion
* Default packages are not allowed. It can cause particular problems for Spring Boot applications that use the ``@ComponentScan``, ``@EntityScan`` or ``@SpringBootApplication`` annotations since every class from every jar is read.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+============
+
+* Write performant queries that can also deal with more 1000 objects in a reasonable time
+* Prefer one query that fetches additional data instead of many small queries, but don't overdo it. A good rule of thumb is to query not more than 3 associations at the same time.
```
```suggestion
* Prefer one query that fetches additional data instead of many small queries, but don't overdo it. A good rule of thumb is to query no more than 3 associations at the same time.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+
+* Write performant queries that can also deal with more 1000 objects in a reasonable time
+* Prefer one query that fetches additional data instead of many small queries, but don't overdo it. A good rule of thumb is to query not more than 3 associations at the same time.
+* Think about lazy vs. eager fetching when modeling the data types
```
```suggestion
* Think about lazy vs. eager fetching when modeling the data types.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* Write performant queries that can also deal with more 1000 objects in a reasonable time
+* Prefer one query that fetches additional data instead of many small queries, but don't overdo it. A good rule of thumb is to query not more than 3 associations at the same time.
+* Think about lazy vs. eager fetching when modeling the data types
+* Simple datatypes: immediately think about whether null should be supported as additional state or not. In most cases it is preferable to avoid null
```
```suggestion
* Simple datatypes: immediately think about whether null should be supported as additional state or not. In most cases it is preferable to avoid null.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* Prefer one query that fetches additional data instead of many small queries, but don't overdo it. A good rule of thumb is to query not more than 3 associations at the same time.
+* Think about lazy vs. eager fetching when modeling the data types
+* Simple datatypes: immediately think about whether null should be supported as additional state or not. In most cases it is preferable to avoid null
+* Use ``Timestamp`` instead of ``Datetime``
```
```suggestion
* Use ``Timestamp`` instead of ``Datetime``.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+8. Comments:
+============
+
+Only if we are writing complicated algorithm and our comments will help other developers to better understand it. If we are adding a comment then our code is not self-explaining.
```
```suggestion
Only write comments for complicated algorithms to help other developers better understand them. If we add a comment, our code is not self-explanatory.
```

sjagla27 commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+9. Utility:
+===========
+
+Utility methods can and should be placed in a class named for specific functionality, not “miscellaneous stuff related to project”. Most of the time, our static methods belong on a related class.
```
```suggestion
Utility methods can and should be placed in a class named for specific functionality, not “miscellaneous stuff related to project”. Most of the time, our static methods belong in a related class.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+10. Auto configuration:
+=======================
+
+Spring Boot favors Java-based configuration. Although it is possible to use Sprint Boot with XML sources but it’s generally not recommended. You need not put all your @Configuration into a single class. The @Import annotation can be used to import additional configuration classes.
```
```suggestion
Spring Boot favors Java-based configuration. Although it is possible to use Sprint Boot with XML sources but it’s generally not recommended. You need not put all your `@Configuration` into a single class. The `@Import` annotation can be used to import additional configuration classes.
```

anditurdiu replied:
```suggestion
Spring Boot favors Java-based configuration. Although it is possible to use Sprint Boot with XML sources, it is generally not recommended. You need not put all your `@Configuration` into a single class. The `@Import` annotation can be used to import additional configuration classes.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* Don’t write code when you are tired or in a bad mood.
+* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because time and resources that will be spent on hard readable code will be much higher than what you get from optimization.
+* Commit messages should describe both what the commit changes and how it does it.
+* ARCHITECTURE FIRST, writing code without thinking of its architecture is useless in the same way as dreaming about your desires without a plan of achieving them.
```
```suggestion
* ARCHITECTURE FIRST, writing code without thinking of the system's architecture is useless, in the same way as dreaming about your desires without a plan of achieving them.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+===========================
+
+* Always use the least possible access level, prefer using private over public access modifier (protected can be used as well).
+* Previously we used transactions very random, now we have to avoid using Transactional, if not sure that they are needed. Transactions can kill performance, introduce locking issues and database concurrency problems, and add complexity to our application. Good read: https://codete.com/blog/5-common-spring-transactional-pitfalls/
```
```suggestion
* Previously we used transactions very randomly, now we have to avoid using `@Transactional`, if we are not sure that they are needed. Transactions can kill performance, introduce locking issues and database concurrency problems, and add complexity to our application. Good read: https://codete.com/blog/5-common-spring-transactional-pitfalls/
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+
+* Always use the least possible access level, prefer using private over public access modifier (protected can be used as well).
+* Previously we used transactions very random, now we have to avoid using Transactional, if not sure that they are needed. Transactions can kill performance, introduce locking issues and database concurrency problems, and add complexity to our application. Good read: https://codete.com/blog/5-common-spring-transactional-pitfalls/
+* Define a constant if the same value is used more than once. Constants allow to change code later a lot easier, instead of looking for the places where this variable was used, you only need to change it in only one place.
```
```suggestion
* Define a constant if the same value is used more than once. Constants allow you to change code later a lot easier. Instead of looking for the places where this variable was used, you only need to change it in only one place.
```

anditurdiu commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* Always use the least possible access level, prefer using private over public access modifier (protected can be used as well).
+* Previously we used transactions very random, now we have to avoid using Transactional, if not sure that they are needed. Transactions can kill performance, introduce locking issues and database concurrency problems, and add complexity to our application. Good read: https://codete.com/blog/5-common-spring-transactional-pitfalls/
+* Define a constant if the same value is used more than once. Constants allow to change code later a lot easier, instead of looking for the places where this variable was used, you only need to change it in only one place.
+* Facilitate code reuse. Always move duplicated code to reusable methods. Intelij is very good at suggesting duplicated lines and even automatically extracting them. Also don't be shy to use Generics.
```
```suggestion
* Facilitate code reuse. Always move duplicated code to reusable methods. IntelliJ is very good at suggesting duplicated lines and even automatically extracting them. Also don't be shy to use Generics.
```

anditurdiu reviewed (changes requested):
Looks good and reads well. 
Here are some minor improvement suggestions:

sjagla27 reviewed (changes requested):
Reads well. I left some remarks

FuchsDominik commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+12. Dependency injection:
+=========================
+
+* Some of you may argue this, but by favoring constructor injection you can keep your business logic free from Spring. Not only is the @Autowired annotation optional on constructors, you also get the benefit of being able to easily instantiate your bean without Spring.
```
```suggestion
* Some of you may argue this, but by favoring constructor injection you can keep your business logic free from Spring. Not only is the `@Autowired` annotation optional on constructors, you also get the benefit of being able to easily instantiate your bean without Spring.
```

balazs-czopf replied:
```suggestion
* Some of you may argue with this, but by favoring constructor injection you can keep your business logic free from Spring. Not only is the @Autowired annotation optional on constructors, you also get the benefit of being able to easily instantiate your bean without Spring.
```

FuchsDominik reviewed (changes requested):
Looks really good, just found two mini changes.

MaisiKoleni commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* Make variable and methods private by default and increasing access step by step like from a private to protected and not public.
+* Classes, methods or functions should be Open for extension and Closed for modification(Open Closed Design Principle).
+* Program for the interface and not for implementation, you should use interface type on variables, return types of a method or argument type of methods. Just like using SuperClass type to store object rather using SubClass.
+* The use of interface is to facilitate polymorphism, client should not implement an interface method if its not needed.
```
This reads a bit strange, I guess a the/a is missing?

MaisiKoleni commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+* The routes should follow the general structure entity > entityId > sub-entity ... (e.g. "/exercises/{exerciseId}/participations").
+* Use plural for route's entities.
+* Specify the key entity at the end of the route (e.g. "text-editor/participations/{participationId}" should be changed to "participations/{participationId}/text-editor").
+* Never specify an id that is used only for consistency and not used in the code (e.g. GET "/courses/{courseId}/exercises{exerciseId}/participations/{participationId}/submissions/{submissionId}" can be simplified to GET "/submissions/{submissionId}" because all other entities than the submission are either not needed or can be loaded without the need to specify the id).
```
```suggestion
* Never specify an id that is used only for consistency and not used in the code (e.g. GET "/courses/{courseId}/exercises/{exerciseId}/participations/{participationId}/submissions/{submissionId}" can be simplified to GET "/submissions/{submissionId}" because all other entities than the submission are either not needed or can be loaded without the need to specify the id).
```

MaisiKoleni commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+15. General best practices:
+===========================
+
+* Always use the least possible access level, prefer using private over public access modifier (protected can be used as well).
```
```suggestion
* Always use the least possible access level, prefer using private over public access modifier (package-private or protected can be used as well).
```

MaisiKoleni reviewed (changes requested):
See the online comments for some suggestions.

I would also like to add some more Java-related points maybe in a separate PR like using `ZonedDateTime` and `Files` and logging with `{}` instead of string concatenation.

balazs-czopf commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+
+* Always use kebab-case (e.g. "/exampleAssessment" → "/example-assessment").
+* The routes should follow the general structure entity > entityId > sub-entity ... (e.g. "/exercises/{exerciseId}/participations").
+* Use plural for route's entities.
```
```suggestion
* Use plural for a route's entities.
```

balazs-czopf commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+
+* Don’t write complex code.
+* Don’t write code when you are tired or in a bad mood.
+* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because time and resources that will be spent on hard readable code will be much higher than what you get from optimization.
```
```suggestion
* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because the time and resources spent on hard-to-read code cost much more than what we gain through optimization.
```

balazs-czopf commented on docs/dev/guidelines/server.rst:
```diff
@@ -0,0 +1,148 @@
+Spring Boot favors Java-based configuration. Although it is possible to use Sprint Boot with XML sources but it’s generally not recommended. You need not put all your @Configuration into a single class. The @Import annotation can be used to import additional configuration classes.
+One of the flagship features of Spring Boot is its use of Auto-configuration. This is the part of Spring Boot that makes your code simply work. It gets activated when a particular jar file is detected on the classpath. The simplest way to make use of it is to rely on the Spring Boot Starters.
+
+11. Keep your @RestController’s clean and focused:
```
```suggestion
11. Keep your ``@RestController``-s clean and focused:
```

balazs-czopf commented on docs/README.md:
```diff
@@ -1,12 +1,11 @@
 ## Documentation Hosting
 
-The [Artemis documentation] is hosted on [Read the Docs] (RtD) for the `develop` (latest) branch, as well as for git tags. The latest tag is always linked as the as _stable_ version.
+[Read the Docs] (RtD) hosts the [Artemis documentation] for `develop` (latest) branch, as well as for git tags. The latest tag is always the _stable_ version.
```
```suggestion
[Read the Docs] (RtD) hosts the [Artemis documentation] for the `develop` (latest) branch, as well as for git tags. The latest tag is always the _stable_ version.
```

balazs-czopf commented on README.md:
```diff
@@ -11,45 +11,45 @@ Artemis was initially generated using JHipster 6.10.3. ([Documentation and help]
-1. **VCS:** Atlassian Bitbucket Server
-2. **CIS:** Atlassian Bamboo Server
-3. **UMS:** Atlassian JIRA Server (more specifically Atlassian Crowd on the JIRA Server)
+Find here a guide on [how to contribute](/CONTRIBUTING.md) to Artemis.
```
```suggestion
Here you can find a guide on [how to contribute](/CONTRIBUTING.md) to Artemis.
```

balazs-czopf reviewed (changes requested):
Looks pretty good. Left a few inline comments though.
I also noticed an inconsistency when using backticks around @-annotations: there are some with double backticks and some with simple ones.

krusche commented:
Thanks everyone for the feedback. I updated the guidelines accordingly and added client guidelines as well.
Both are still WORK IN PROGRESS, but I guess it's better to have something that nothing.
We can further adapt and improve the guidelines in the upcoming weeks.
